### PR TITLE
Added method to reshape dims of network.

### PR DESCRIPTION
**Pull request by AGaliciaMartinez**
This PR implements two methods, `match_in_dims` and `match_out_dims` that allows splitting edges to match a target dimension. This is primarily meant to be used with `ptrace` as we need to reshape the outer nodes such that the network has the appropriate dimensions.

Note that `_match_edges_by_split` is a very similar function but with different scope. The new `_match_dimensions` is meant to be used with unary operators that required a change in the dimension (ptrace) whereas `_match_edges_by_split` is meant to be used with binary operators (matmul). This is because for  `_match_dimensions` (unary version) we do not allow merging two edge dimension. However, such operation can also be achieved in `_match_edges_by_split` (binary version) by just splitting the edges as two list of edges is available.

The reason to not allow here in `_match_dimensions` to merge edges is because this is a relatively complex operation that will require contracting nodes. I am assuming that it is best to do the contraction of the full network in a single step such that an optimized path is used.

### CODE REVIEW

File: src/qutip_tensornetwork/core/data/network.py [hunk 1]
```diff
@@ -576,6 +576,88 @@ def tensor(self, other):
...
         return Network._fast_constructor(out_edges, in_edges, nodes)
 
+    def match_out_dims(self, target_dims):
+        """Reshape nodes by splitting edges such that `in_edges` matches
+        ``target_dims``. After this function the following will hold:
+            ``network.dims[0] == target_dims``
+
+        Note
+        ----
+        This is an in-place operation that returns itself.
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]: What is the reason for returning self if the operation is always in place?

Aside: It's quite common in, e.g., Javascript to return self to allow operations to be chained using
```javascript
obj
    .red()
    .stripy()
    .invisible()
;
```
But one doesn't see it done often in Python.

File: src/qutip_tensornetwork/core/data/network.py [hunk 2]
```diff
@@ -576,6 +576,88 @@ def tensor(self, other):
...
+        Network
+
+        Examples
+        --------
+        >>>network_dim = [4]
+        >>>target_dims = [2, 2]
+        >>>array = np.random.random(network_dim)
+        >>>node = tn.Node(array)
+        >>>network = Network(node[:], [])
+        >>>network.match_out_dims(target_dims)
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
        >>> network_dim = [4]
        >>> target_dims = [2, 2]
        >>> array = np.random.random(network_dim)
        >>> node = tn.Node(array)
        >>> network = Network(node[:], [])
        >>> network.match_out_dims(target_dims)
```

File: src/qutip_tensornetwork/core/data/network.py [hunk 3]
```diff
@@ -576,6 +576,88 @@ def tensor(self, other):
...
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
        >>> network_dim = [4]
        >>> target_dims = [2, 2]
        >>> array = np.random.random(network_dim)
        >>> node = tn.Node(array)
        >>> network = Network(node[:], [])
        >>> network.match_out_dims(target_dims)
```

File: src/qutip_tensornetwork/core/data/network.py [hunk 4]
```diff
@@ -672,3 +754,108 @@ def _match_edges_by_split(out_edges, in_edges):
...
+    function is private as it is meant to be used only within network.
+
+    Examples
+    --------
+    >>>network_dim = [4]
+    >>>target_dims = [2, 2]
+    >>>array = np.random.random(network_dim)
+    >>>node = tn.Node(array)
+    >>>network = Network(node[:], [])
+    >>>_match_dimensions(network.in_edges, target_dims)
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
    >>> network_dim = [4]
    >>> target_dims = [2, 2]
    >>> array = np.random.random(network_dim)
    >>> node = tn.Node(array)
    >>> network = Network(node[:], [])
    >>> _match_dimensions(network.in_edges, target_dims)
```

File: src/qutip_tensornetwork/core/data/network.py [hunk 5]
```diff
@@ -672,3 +754,108 @@ def _match_edges_by_split(out_edges, in_edges):
...
+    ValueError
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]: Hmm. Should this first example be returning a ValueError?

File: src/qutip_tensornetwork/core/data/network.py [hunk 6]
```diff
@@ -672,3 +754,108 @@ def _match_edges_by_split(out_edges, in_edges):
...
+
+    Since only splitting of edges are allowed, the next example will raise a
+    ValueError.
+    >>>network_dim = [2, 2]
+    >>>target_dims = [4]
+    >>>array = np.random.random(network_dim)
+    >>>node = tn.Node(array)
+    >>>network = Network(node[:], [])
+    >>>_match_dimensions(network.in_edges, target_dims)
+    ValueError
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
    Since only splitting of edges are allowed, the next example will raise a
    ValueError:
    
    >>> network_dim = [2, 2]
    >>> target_dims = [4]
    >>> array = np.random.random(network_dim)
    >>> node = tn.Node(array)
    >>> network = Network(node[:], [])
    >>> _match_dimensions(network.in_edges, target_dims)
    ValueError
```

File: src/qutip_tensornetwork/core/data/network.py [hunk 7]
```diff
@@ -672,3 +754,108 @@ def _match_edges_by_split(out_edges, in_edges):
...
+    """
+
+    edges = edges[:]
+    target_dims = target_dims[:]
+    _target_dims = target_dims[:]  # This is kept for the error messages.
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]: I would do the input checks (the `if len(edges) ...` etc) first and then construct these temporary values afterwards so they're constructed closer to the loop they're used in.

I would also give the copies new names (e.g. `edges_remaining`, `dims_remaining`) rather than overwriting the original names and creating `target_dims2`.

In the case of `edges`, I would also be in favour of replacing the contents of the `edges` list at the end. Currently at the end of the function the passed in list is empty and the original edges within the list are a modified, so we might as well put the updates edges back in the original list.

File: src/qutip_tensornetwork/core/data/network.py [hunk 8]
```diff
@@ -672,3 +754,108 @@ def _match_edges_by_split(out_edges, in_edges):
...
+
+    e_dims = [e.dimension for e in edges]
+
+    new_edges = []
+
+    if len(edges) == 0 and len(target_dims) == 0:
+        return _edges
+
+    if len(edges) == 0 or len(target_dims) == 0:
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
    if len(edges) == 0 or len(target_dims) == 0 or np.prod(e_dims) != np.prod(target_dims):
```
This avoids repeating the exception below and the `or` will short circuit if edges or target_dims are empty.

File: src/qutip_tensornetwork/core/data/network.py [hunk 9]
```diff
@@ -672,3 +754,108 @@ def _match_edges_by_split(out_edges, in_edges):
...
+        )
+
+    if np.prod(e_dims) != np.prod(target_dims):
+        raise ValueError(
+            "edges are not compatible. The dimensions of edges is "
+            + str(e_dims)
+            + ", whereas the target dimension is "
+            + str(_target_dims)
+        )
+
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
```
Can be removed if suggestion above is applied.

File: src/qutip_tensornetwork/core/data/network.py [hunk 10]
```diff
@@ -672,3 +754,108 @@ def _match_edges_by_split(out_edges, in_edges):
...
+            new_edges.append(edge)
+
+            if len(edges) == 0 and len(target_dims) == 0:
+                break
+
+            edge = edges.pop()
+            target_dim = target_dims.pop()
+
+        elif edge.dimension > target_dim:  # Split edge
+            if edge.dimension % target_dim != 0:
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py:820]: When `ValueError` is raised here the edge list is a mess. This means someone could call `qobj.ptrace()` and then when an exception is raised the Network referenced by `qobj._data` is in an invalid state. There are two good options -- copy all the edges at the start (which has some downsides) or check that everything will succeed at the start (probably with a similar loop that doesn't call `tn.split_edge`).

With the second approach, one can probably keep a record of the needed edges and splits and just apply those in a simpler loop afterwards.

File: tests/core/data/test_network.py [hunk 11]
```diff
@@ -352,6 +353,80 @@ def test_from_2d_array(shape, expected_dims):
...
+        ([4], [2, 2]),
+        ([8], [2, 2, 2]),
+        ([2, 2], [2, 2]),
+        ([2, 4, 2], [2, 2, 2, 2]),
+        ([2, 15, 7], [2, 3, 5, 7]),
+    ],
+)
+class TestMatchDims:
+    """This class contians all the tests for the functions that change the
+    dimsnension of a network by splitting edges.
```
Comment (hodgestar) [inline tests/core/data/test_network.py]:
```suggestion
    """This class contains all the tests for the functions that change the
    dimensions of a network by splitting edges.
```

File: tests/core/data/test_network.py [hunk 12]
```diff
@@ -352,6 +353,80 @@ def test_from_2d_array(shape, expected_dims):
...
+
+@pytest.mark.parametrize(
+    "dim_edges, target_dims",
+    [
+        ([2, 2], [4]),
+        ([2, 2], [3]),
+        ([2, 3, 5], [2, 5, 3]),
+    ],
+)
+class TestMatchDimsRaisise:
```
Comment (hodgestar) [inline tests/core/data/test_network.py]:
```suggestion
class TestMatchDimsRaise:
```

File: tests/core/data/test_network.py [hunk 13]
```diff
@@ -352,6 +353,80 @@ def test_from_2d_array(shape, expected_dims):
...
+    """This class contians the tests for the functions that change the
+    dimsnension of a network by splitting edges. In particular it ensures that
+    thep proper error is raised.
+    """
```
Comment (hodgestar) [inline tests/core/data/test_network.py]:
```suggestion
    """This class contains the tests for the functions that change the
    dimensions of a network by splitting edges. In particular it ensures that
    the correct errors are raised.
    """
```

File: tests/core/data/test_network.py [hunk 14]
```diff
@@ -352,6 +353,80 @@ def test_from_2d_array(shape, expected_dims):
...
+        node = random_node(dim_edges)
+        network = Network(node[:], [], copy=False)
+        with pytest.raises(ValueError):
+            _match_dimensions(network.out_edges, target_dims)
+
+    def test_match_out_dims_raises(self, dim_edges, target_dims):
+        node = random_node(dim_edges)
+        network = Network(node[:], [], copy=False)
+        with pytest.raises(ValueError):
+            network.match_out_dims(target_dims)
```
Comment (hodgestar) [inline tests/core/data/test_network.py:432]: It's probably worthwhile asserting that `network` is in a valid state after an error has been raised.

**Review by hodgestar — CHANGES_REQUESTED** [review]
The approach looks good, but I left some questions about the handling of the error cases.

File: src/qutip_tensornetwork/core/data/network.py (hunk 5 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensornetwork/core/data/network.py]: It should not. Thanks for catching it.

File: tests/core/data/test_network.py (hunk 14 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py:432]: That's a good  point. I will do so.

**Comment by AGaliciaMartinez** [comment]
Thank you for the review Simon!

File: src/qutip_tensornetwork/core/data/network.py (hunk 1 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensornetwork/core/data/network.py]: I do not have any strong reason to make it return self. I do like the fact that the following code:
```python
network.match_out_dims() 
```
will print the network. I can remove it if you believe it can cause any problems.

Comment (hodgestar) [reply src/qutip_tensornetwork/core/data/network.py]: I don't think it will directly cause problems, I just think it will surprise Python people a bit that `network.match_out/in_dims(...)` will both modify `network` and return it. :)

File: src/qutip_tensornetwork/core/data/network.py (hunk 10 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensornetwork/core/data/network.py:820]: I am going to go with the copy approach just because it is simpler. We may want to revisit this in the future though.

**Comment by AGaliciaMartinez** [comment]
I changed so that `match_in_dims` and `match_out_dims` return a new Network instead of performing an inplace modification. @hodgestar, could you take a look at the latest changes?

File: src/qutip_tensornetwork/core/data/network.py [hunk 15]
```diff
@@ -576,6 +576,82 @@ def tensor(self, other):
 
         return Network._fast_constructor(out_edges, in_edges, nodes)
 
+    def match_out_dims(self, target_dims):
+        """Reshape nodes by splitting edges such that `in_edges` matches
+        ``target_dims``. After this function the following will hold:
+            ``network.dims[0] == target_dims``
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py:582]:
```suggestion
        ``target_dims``. The returned network will have ``network.dims[0] == target_dims``.
```

File: src/qutip_tensornetwork/core/data/network.py [hunk 16]
```diff
@@ -576,6 +576,82 @@ def tensor(self, other):
...
+        """
+        out = self.copy()
+        edges = _match_dimensions(out.out_edges, target_dims)
+        out.out_edges = edges
+        return out
+
+    def match_in_dims(self, target_dims):
+        """Reshape nodes by splitting edges such that `in_edges` matches
+        ``target_dims``. After this function the following will hold:
+            ``network.dims[1] == target_dims``
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py:620]:
```suggestion
        ``target_dims``. The returned network will have ``network.dims[1] == target_dims``.
```

File: src/qutip_tensornetwork/core/data/network.py [hunk 17]
```diff
@@ -576,6 +576,82 @@ def tensor(self, other):
...
+
+        Parameters
+        ----------
+        target_dims: list of int
+            Desired dimensions for ``out_edges``.
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py:625]:
```suggestion
            Desired dimensions for ``in_edges``.
```

File: src/qutip_tensornetwork/core/data/network.py [hunk 18]
```diff
@@ -576,6 +576,82 @@ def tensor(self, other):
...
+        Network
+
+        Examples
+        --------
+        >>> network_dim = [4]
+        >>> target_dims = [2, 2]
+        >>> array = np.random.random(network_dim)
+        >>> node = tn.Node(array)
+        >>> network = Network(node[:], [])
+        >>> network = network.match_out_dims(target_dims)
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py:638]:
```suggestion
        >>> network = network.match_in_dims(target_dims)
```

File: src/qutip_tensornetwork/core/data/network.py [hunk 19]
```diff
@@ -672,3 +748,103 @@ def _match_edges_by_split(out_edges, in_edges):
...
+
+    target_dims = target_dims[:]
+    _target_dims = target_dims[:]  # This is kept for the error messages.
+
+    e_dims = [e.dimension for e in edges]
+
+    new_edges = []
+
+    if len(edges) == 0 and len(target_dims) == 0:
+        return _edges
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py:792]: Should this just be `return edges`? `_edges` does not seem to be defined any more. We should add a test for the empty edges and target_dims case.

File: src/qutip_tensornetwork/core/data/network.py [hunk 20]
```diff
@@ -672,3 +748,103 @@ def _match_edges_by_split(out_edges, in_edges):
...
+    if (
+        len(edges) == 0
+        or len(target_dims) == 0
+        or np.prod(e_dims) != np.prod(target_dims)
+    ):
+        raise ValueError(
+            "edges are not compatible. The dimensions for edges is "
+            + str(e_dims)
+            + ", whereas the target dimension is"
+            + str(_target_dims)
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py:803]:
```suggestion
            f"edges are not compatible. The dimensions for edges is {e_dims}"
            f", whereas the target dimension is {_target_dims}."
```

Comment (hodgestar) [reply src/qutip_tensornetwork/core/data/network.py:803]: I guess we should just use f-strings these days.

File: src/qutip_tensornetwork/core/data/network.py [hunk 21]
```diff
@@ -672,3 +748,103 @@ def _match_edges_by_split(out_edges, in_edges):
...
+    """
+
+    edges = edges[:]
+
+    target_dims = target_dims[:]
+    _target_dims = target_dims[:]  # This is kept for the error messages.
+
+    e_dims = [e.dimension for e in edges]
+
+    new_edges = []
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py:789]: I would move the initialization of `new_edges` to just before the top of the loop so that it's closer to where it's used. Currently we define it, then don't use it for awhile, then suddenly use it in the middle of a loop.

**Review by hodgestar — CHANGES_REQUESTED** [review]
Left a couple more suggestions. Thank you for sorting out the state of the network post errors by copying the network -- I think that was a good solution.
